### DISCUSSION TITLE
Call history?

### DISCUSSION THREAD

**Discussion by CharlesHe16** [guidance-ai/guidance#119] [General]
Is there an easy way to access the call history? 

I can see API calls, e.g in `_openai.py` and I could save `kwargs`, but is there a built-in way to do this?

**Comment by slundberg** [comment]
Better logging is something that still needs to happen. So if you have a wish list for the API feel free to share.

**Comment by CharlesHe16** [comment]
What I need to see is how tokens are used, which helps manage things like client-side throttling. It would also be helpful in evaluating how clock time, tokens, and calls compare across different strategies (e.g. since most generation methods involve sending multiple calls under the hood).

I'm not entirely sure, but it seems like the `executed guidance program` just needs to save the `kwargs` input, and the output of the calls. I think a lot of this is already included in `_openai.py` but not saved.

I'm considering using a monkey path or something similar to accomplish this?

While Guidance looks fantastic, I'm not sure how my modification would fit in. 

Not sure how to further engage here.
